src/charm.py

```diff
@@ -52,6 +98,56 @@ def _install_upf_snap(self) -> None:
             logger.error("An exception occurred when installing the UPF snap. Reason: %s", str(e))
             raise e
 
+    def _generate_upf_config_file(self) -> None:
+        """Generate the UPF configuration file."""
+        core_ip_address = self._get_core_network_ip_config()
+        content = render_upf_config_file(
+            upf_hostname=self._get_upf_hostname(),
+            upf_mode=self._get_upf_mode(),
+            access_interface_name=UPF_ACCESS_INTERFACE_NAME,
+            core_interface_name=UPF_CORE_INTERFACE_NAME,
+            core_ip_address=core_ip_address.split("/")[0] if core_ip_address else "",
+            dnn=self._get_dnn_config(),
+            pod_share_path=UPF_CONFIG_PATH,
+            enable_hw_checksum=self._get_enable_hw_checksum(),
+        )
+        if not self._upf_config_file_is_written() or not self._upf_config_file_content_matches(
+            content=content
+        ):
+            self._write_upf_config_file(content=content)
+
+    def _upf_config_file_is_written(self) -> bool:
+        """Return whether the UPF config file was written to the workload."""
+        return self._machine.exists(path=f"{UPF_CONFIG_PATH}/{UPF_CONFIG_FILE_NAME}")
+
+    def _upf_config_file_content_matches(self, content: str) -> bool:
+        """Return whether the UPF config file content matches the provided content."""
+        existing_content = self._machine.pull(path=f"{UPF_CONFIG_PATH}/{UPF_CONFIG_FILE_NAME}")
+        try:
+            return json.loads(existing_content) == json.loads(content)
+        except json.JSONDecodeError:
+            return False
+
+    def _write_upf_config_file(self, content: str) -> None:
+        """Write the UPF config file to the workload."""
+        self._machine.push(path=f"{UPF_CONFIG_PATH}/{UPF_CONFIG_FILE_NAME}", source=content)
+        logger.info("Pushed %s config file", UPF_CONFIG_FILE_NAME)
+
+    def _get_core_network_ip_config(self) -> str:
+        return "192.168.250.3/24"
+
+    def _get_upf_hostname(self) -> str:
+        return "0.0.0.0"
+
+    def _get_upf_mode(self) -> str:
+        return "af_packet"
+
+    def _get_dnn_config(self) -> str:
+        return "internet"
+
+    def _get_enable_hw_checksum(self) -> bool:
+        return True
```

Comment:
I think we should introduce the configs with the defaults in the right place instead of hardcoding the defaults here. We could omit handling `config_changed` for now however.

Reply:
I just added the config.yaml file with some minimal config validation. Currently the file only contains the values that we can interchange without any logic change (Ex. interface name, dnn). I kept out the config requiring more code changes (Ex. upf mode) so that this gets added in a PR of its own. I don't want to add the option to change the upf mode if the code does not support it. Let me know what you think.